examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+    Vector<double> utmp(src.block(0));
+
+    // First solve with the approximation for S
+    {
+      SolverControl solver_control(1000, 1e-6 * src.block(1).l2_norm());
+      SolverCG<>    cg (solver_control);
+
+      dst.block(1) = 0.0;
+      cg.solve(schur_complement_matrix,
+               dst.block(1),
+               src.block(1),
+               preconditioner_S);
+
+      n_iterations_S += solver_control.last_step();
+      dst.block(1) *= -1.0;
+    }
+
+    // Second, apply the top right block (B^T)
+    {
+      system_matrix.block(0,1).vmult(utmp, dst.block(1));
+      utmp *= -1.0;
+      utmp += src.block(0);
+    }
+
+    // Finally, either solve with the top left block (if do_solve_A==true)
+    // or just apply one preconditioner sweep
+    if (do_solve_A == true)
+      {
+        SolverControl solver_control(10000, utmp.l2_norm()*1e-2, true);
+        SolverCG<>    cg (solver_control);
+
+        dst.block(0) = 0.0;
+        cg.solve(system_matrix.block(0,0),
+                 dst.block(0),
+                 utmp,
+                 preconditioner_A);
+
+        n_iterations_A += solver_control.last_step();
+      }
+    else
+      {
+        preconditioner_A.vmult (dst.block(0), utmp);
+        n_iterations_A += 1;
+      }
+  }
+
+  // @sect3{The StokesProblem class}
+  //
+  // This is the main class of the problem.
+  template <int dim>
+  class StokesProblem
+  {
+  public:
+    StokesProblem (const unsigned int degree, SolverType::type solver_type);
+    void run ();
+
+  private:
+    void setup_dofs ();
+    void assemble_system ();
+    void assemble_multigrid ();
+    void solve ();
+    void compute_errors ();
+    void output_results (const unsigned int refinement_cycle) const;
+
+    const unsigned int   degree;
+    SolverType::type     solver_type;
+
+    Triangulation<dim>   triangulation;
+    FESystem<dim>        fe;
+    FESystem<dim>        velocity_fe;
+    DoFHandler<dim>      dof_handler;
+    DoFHandler<dim>      velocity_dof_handler;
+
+    ConstraintMatrix     constraints;
+
+    BlockSparsityPattern      sparsity_pattern;
+    BlockSparseMatrix<double> system_matrix;
+    SparseMatrix<double> pressure_mass_matrix;
+
+    BlockVector<double> solution;
+    BlockVector<double> system_rhs;
+
+    MGLevelObject<SparsityPattern>        mg_sparsity_patterns;
+    MGLevelObject<SparseMatrix<double> >  mg_matrices;
+    MGLevelObject<SparseMatrix<double> >  mg_interface_matrices;
+    MGConstrainedDoFs                     mg_constrained_dofs;
+
+    TimerOutput computing_timer;
+  };
+
+
+
+  template <int dim>
+  StokesProblem<dim>::StokesProblem (const unsigned int degree, SolverType::type solver_type)
+    :
+    degree (degree),
+    solver_type (solver_type),
+    triangulation (Triangulation<dim>::maximum_smoothing),
+    fe (FE_Q<dim>(degree+1), dim, // Finite element for whole system
+        FE_Q<dim>(degree), 1),
+    velocity_fe (FE_Q<dim>(degree+1), dim), // Finite element for velocity-only
```

Comment:
Can you try to avoid the (error prone) code duplication by first initializing `velocity_fe` and then initializing `fe` as
```
  fe (velocity_fe, 1,
       FE_Q<dim> (degree), 1),
```
? You'll also need to swap the order of the declarations of these two variables in the class declaration above to make this work.